docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,324 @@
+
+This would be a huge improvement over the “standard” `kubectl drain` aligned implementation we have today and help to 
+solve a family of issues identified when running Cluster API in production.
+
+## Motivation
+
+While the “standard” `kubectl drain` rules have served us well, new user stories have emerged where a more sophisticated 
+behavior is needed.
+
+### User Stories
+
+**Story 1: Define drain order**
+
+As a cluster admin/user, I would like to define in which order Pods are evicted. For example, I want to configure that 
+Portworx Pods are evicted after all other Pods ([CAPI-11024](https://github.com/kubernetes-sigs/cluster-api/issues/11024)).
+
+Note:  As of today the only way to address this use case with Cluster API is to implement a pre-drain hook with a custom 
+drain process. Due to the complexity of this solution it is not viable for most users.
+
+**Story 2: Exclude “undrainable” Pods from drain**
+
+As a cluster admin/user, I would like the node drain process to ignore pods that will obviously not drain properly 
+(e.g., because they're tolerating the unschedulable or all taints).
+
+Note: As of today Pods that are tolerating the `node.kubernetes.io/unschedulable:NoSchedule` taint will just get rescheduled 
+to the Node after they have been evicted, and this can lead to race conditions and then prevent the Machine to be deleted.
+
+**Story 3: Exclude Pods that should not be drained from drain**
+
+As a cluster admin/user, I would like to instruct the Node drain process to exclude some Pods, e.g. because I don't 
+care if they're gracefully terminated and they should run until the Node goes offline (e.g. monitoring agents).
+
+**Story 4: Define drain order & exclude Pods from drain without management cluster access**
+
+As a workload cluster admin/user, I want to be able to define drain order of Pods and also exclude Pods from drain 
+(similar to Story 1-3, just without management cluster access).
+
+**Story 5: Apply drain configuration to a set of Machines / Clusters**
+
+As a cluster admin/user, I would like to be able to configure the drain behavior for all Machines of a management 
+cluster or for a subset of these Machines without having to configure each Machine individually. I would also like 
+to be able to change the drain configuration without having to modify all Machine objects.
+
+### Goals
+
+* Allow users to customize how/if Pods are evicted:
+  * Allow to define the order in which Pods are evicted
+  * Allow to define if Pods should be excluded from drain
+* Use the current drain default rules as a fallback if neither `MachineDrainRule` CRs nor drain annotations
+  apply to a Pod (this also ensures compatibility with current behavior)
+* Stop waiting for volume detachment for Pods that are excluded from drain
+
+### Non-Goals
+
+* Change the drain behavior for DaemonSet and static Pods (we’ll continue to skip draining for both)
```

Comment:
if we automatically skip volume detachement for daemonSets is a behavioural change probably worth mention in the # Upgrade Strategy section as well

Reply:
Added a not here that the volume detach behavior will change. Same in upgrade strategy